File: java/binary_tree_level_order_traversal_ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res=new ArrayList<List<Integer>>();
        if (root==null) return res;
        ArrayDeque<TreeNode> queue=new ArrayDeque<>();
        int currLevel=1, nextLevel=0;
        List<Integer> level=new ArrayList<>();
        queue.offer(root);
    
        while (!queue.isEmpty()) {
            TreeNode f=queue.poll();
            level.add(f.val);
            currLevel--;
            
            if (f.left!=null) {
                queue.offer(f.left);
                nextLevel++;
            }
            if (f.right!=null) {
                queue.offer(f.right);
                nextLevel++;
            }
            
            if (currLevel==0) {
                res.add(new ArrayList(level));
                level.clear();
                currLevel=nextLevel;
                nextLevel=0;
            }
        }
    
        Collections.reverse(res);
        return res;
    }
}

File: cpp/binary-tree-level-order-traversal-ii.cpp
107. Binary Tree Level Order Traversal II Add to List
DescriptionSubmissionsSolutions
Total Accepted: 120136
Total Submissions: 309232
Difficulty: Easy
Contributor: LeetCode
Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).

For example:
Given binary tree [3,9,20,null,null,15,7],
    3
   / \
  9  20
    /  \
   15   7
return its bottom-up level order traversal as:
[
  [15,7],
  [9,20],
  [3]
]
Subscribe to see which companies asked this question.

Hide Tags Tree Breadth-first Search
Hide Similar Problems (M) Binary Tree Level Order Traversal

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> res;
        if (root==NULL) return res;
        queue<TreeNode*> q;
        q.push(root);
        int currNum;
        while (!q.empty()) {
            currNum=q.size();
            vector<int> level;
            for (int i=0; i<currNum; i++) {
                TreeNode* f=q.front();
                q.pop();
                level.push_back(f->val);
                if (f->left) q.push(f->left);
                if (f->right) q.push(f->right);
            }
            res.push_back(level);
        }
        
        return vector<vector<int>>(res.rbegin(), res.rend());
    }
};

